agent: Allow writing CNI configuration when ready

tgraf opened the pull request:
This can be used to generate the CNI configuration in a location outside of the
cni-conf-dir and then copy the CNI configuration into the config directory once
the agent is up. This avoids the node being marked ready before the agent is
ready to schedule pods.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8316)
<!-- Reviewable:end -->

nebril commented on daemon/daemon_main.go:
```diff
@@ -769,6 +769,9 @@ func init() {
 	flags.Bool(option.SkipCRDCreation, false, "Skip Kubernetes Custom Resource Definitions creations")
 	option.BindEnv(option.SkipCRDCreation)
 
+	flags.String(option.WriteCNIConfigurationWhenReady, "", "Write the CNI configuration to path when agent is ready")
```
Can we add an information for user that `ReadCNIConfiguration` also needs to be defined for this to work?

tgraf replied:
Added.

nebril commented on daemon/daemon_main.go:
```diff
@@ -1371,6 +1374,18 @@ func runDaemon() {
 	log.WithField("bootstrapTime", time.Since(bootstrapTimestamp)).
 		Info("Daemon initialization completed")
 
+	if option.Config.WriteCNIConfigurationWhenReady != "" {
```
I think we should check if `option.Config.ReadCNIConfiguration` is not empty here and let user know with a distinct error message that it should be fixed.

ianvernon replied:
This is already done by the check in `DaemonConfig.Validate()` below - we check if both are set and return an error if so. The call to `Validate()` for said configuration is done before we hit this code, so we'd never hit this point if only one option is set.

tgraf committed:
agent: Allow writing CNI configuration when ready

This can be used to generate the CNI configuration in a location outside of the
cni-conf-dir and then copy the CNI configuration into the config directory once
the agent is up. This avoids the node being marked ready before the agent is
ready to schedule pods.

Signed-off-by: Thomas Graf <thomas@cilium.io>
